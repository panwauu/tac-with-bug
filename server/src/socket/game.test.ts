import type { GameSocketC } from '../../../shared/types/GameNamespaceDefinition';

import { TacServer } from '../server';
import supertest from 'supertest';
import { registerNUsersWithSockets, unregisterUsersWithSockets, userWithCredentialsAndSocket, registerGameSocket, unregisterGameSocket } from '../helpers/userHelper';

describe('Game Test Suite via Socket.io', () => {
    let usersWithSockets: userWithCredentialsAndSocket[], agent: supertest.SuperAgentTest, server: TacServer, gameBefore: any;
    const gameID = 96;
    const gameOskar = { playerIndex: 2, userid: 4 }
    const gameSophia = { playerIndex: 3, userid: 7 }

    beforeAll(async () => {
        server = new TacServer()
        await server.listen(1234)
        agent = supertest.agent(server.httpServer)
        usersWithSockets = await registerNUsersWithSockets(server, agent, 3);
        await server.pgPool.query('UPDATE users_to_games SET userid = $1 WHERE player_index = $2 AND gameid = $3;', [usersWithSockets[0].id, gameOskar.playerIndex, gameID])
        await server.pgPool.query('UPDATE users_to_games SET userid = $1 WHERE player_index = $2 AND gameid = $3;', [usersWithSockets[1].id, gameSophia.playerIndex, gameID])
        await server.pgPool.query('UPDATE games SET status=\'running\' WHERE id = $1;', [gameID])
        gameBefore = await server.pgPool.query('SELECT game FROM games WHERE id = $1;', [gameID]).then((r) => r.rows[0].game)
    })

    afterAll(async () => {
        await server.pgPool.query('UPDATE users_to_games SET userid = $1 WHERE player_index = $2 AND gameid = $3;', [gameOskar.userid, gameOskar.playerIndex, gameID])
        await server.pgPool.query('UPDATE users_to_games SET userid = $1 WHERE player_index = $2 AND gameid = $3;', [gameSophia.userid, gameSophia.playerIndex, gameID])
        await server.pgPool.query('UPDATE games SET status=\'aborted\', game=$2, rematch_open=false WHERE id = $1;', [gameID, JSON.stringify(gameBefore)])
        await unregisterUsersWithSockets(agent, usersWithSockets)
        await server.destroy()
    })

    describe('Test invalid connection', () => {
        let gameSocket: GameSocketC;

        afterEach(async () => {
            await unregisterGameSocket(gameSocket)
        })

        test('Test with invalid game', async () => {
            await expect(registerGameSocket('test', usersWithSockets[0].token)).rejects.toBe(undefined)
        })

        test('Test with invalid token', async () => {
            await expect(registerGameSocket(1000, 'test')).rejects.toBe(undefined)
        })
    })

    describe('Test valid connection', () => {
        let gameSocket: GameSocketC, anotherGameSocket: GameSocketC;

        afterEach(async () => {
            await unregisterGameSocket(gameSocket)
            await unregisterGameSocket(anotherGameSocket)
        })

        test('Test with own game', async () => {
            gameSocket = await registerGameSocket(gameID, usersWithSockets[0].token)
        })

        test('Test with other tournament game', async () => {
            gameSocket = await registerGameSocket(2660, usersWithSockets[0].token)
        })
    })

    /*describe('Test dealCards', () => {
        let gameSocket: GameSocketC, interval: any;

        beforeAll(async () => {
            const gameCopy = cloneDeep(gameBefore)
            gameCopy.cardsWithMoves = []
            gameCopy.cards.players.forEach((_: any, i: number) => gameCopy.cards.players[i] = [])
            await server.pgPool.query('UPDATE games SET game=$2 WHERE id = $1;', [gameID, JSON.stringify(gameCopy)])
            await new Promise((resolve) => setTimeout(() => resolve(null), 200)) // Needed to fix test
        })

        afterAll(async () => {
            clearInterval(interval);
            await unregisterGameSocket(gameSocket);
        })

        test('Register player and expect dealCards', async () => {
            gameSocket = initiateGameSocket(gameID, usersWithSockets[0].token)
            let nUpdates = 0;
            let updateData: any = null;
            gameSocket.on('update', (data) => { updateData = data; nUpdates += 1 })


            const promiseArray = [
                new Promise<any>((resolve) => gameSocket.once('game:online-players', (data) => resolve(data))),
                new Promise<any>((resolve) => interval = setInterval(() => {
                    if (nUpdates >= 2) {
                        resolve(updateData);
                        clearInterval(interval)
                    }
                }, 20)),
            ]
            await waitForGameSocketConnection(gameSocket)
            const result = await Promise.all(promiseArray)
            expect(result[0].onlineGamePlayers).toEqual([gameOskar.playerIndex])
            expect(result[0].nWatchingPlayers).toEqual(0)
            expect(result[0].watchingPlayerNames).toEqual([])
            expect(generateGameSnapshot(result[1])).toMatchSnapshot()
            expect(result[1].ownCards.length).toBeGreaterThan(0)
        })
    })

    describe('Test complete flow with all events', () => {
        let gameSocket0: GameSocketC, gameSocket1: GameSocketC;

        afterAll(async () => {
            await unregisterGameSocket(gameSocket0)
            await unregisterGameSocket(gameSocket1)
        })

        test('Register first player', async () => {
            await new Promise<void>((resolve) => { setTimeout(() => { resolve() }, 500) })
            gameSocket0 = initiateGameSocket(gameID, usersWithSockets[0].token)
            const promiseArray = [
                new Promise<any>((resolve) => gameSocket0.once('game:online-players', (data) => resolve(data))),
                new Promise<any>((resolve) => gameSocket0.once('update', (data) => resolve(data))),
            ]
            await waitForGameSocketConnection(gameSocket0)
            const result = await Promise.all(promiseArray)
            expect(result[0].onlineGamePlayers).toEqual([gameOskar.playerIndex])
            expect(result[0].nWatchingPlayers).toEqual(0)
            expect(result[0].watchingPlayerNames).toEqual([])
            expect(generateGameSnapshot(result[1])).toMatchSnapshot();
        })

        test('Register second player', async () => {
            gameSocket1 = initiateGameSocket(gameID, usersWithSockets[1].token)
            const promiseArray = [
                new Promise<any>((resolve) => gameSocket0.once('game:online-players', (data) => resolve(data))),
                new Promise<any>((resolve) => gameSocket1.once('game:online-players', (data) => resolve(data))),
                new Promise<any>((resolve) => gameSocket1.once('update', (data) => resolve(data))),
            ]
            await waitForGameSocketConnection(gameSocket1)
            const result = await Promise.all(promiseArray)
            expect(result[0].onlineGamePlayers.sort()).toEqual([gameOskar.playerIndex, gameSophia.playerIndex].sort())
            expect(result[0].nWatchingPlayers).toEqual(0)
            expect(result[0].watchingPlayerNames).toEqual([])
            expect(result[1].onlineGamePlayers.sort()).toEqual([gameOskar.playerIndex, gameSophia.playerIndex].sort())
            expect(result[1].nWatchingPlayers).toEqual(0)
            expect(result[1].watchingPlayerNames).toEqual([])
            expect(generateGameSnapshot(result[2])).toMatchSnapshot();
        })

        test('Post move', async () => {
            const promiseArray = [
                new Promise<any>((resolve) => gameSocket0.once('update', (data) => resolve(data))),
                new Promise<any>((resolve) => gameSocket1.once('update', (data) => resolve(data))),
            ]
            gameSocket1.emit('postMove', [3, 0, 'beenden'])
            const result = await Promise.all(promiseArray)
            expect(generateGameSnapshot(result[0])).toMatchSnapshot();
            expect(generateGameSnapshot(result[1])).toMatchSnapshot();
            const game = result[0]
            expect(game.status).toBe('won-0')
            expect(game.rematch_open).toBe(true)
        })

        test('Does emit onlinePlayers to gameSocket1 when gameSocket0 disconnects', async () => {
            const promiseArray = [
                new Promise<any>((resolve) => gameSocket1.once('game:online-players', (data) => resolve(data))),
            ]
            gameSocket0.disconnect()
            const result = await Promise.all(promiseArray);
            expect(result[0].onlineGamePlayers).toEqual([gameSophia.playerIndex])
            expect(result[0].nWatchingPlayers).toEqual(0)
            expect(result[0].watchingPlayerNames).toEqual([])
        })

        test('Unregister last player', async () => {
            gameSocket1.disconnect()
        })
    })*/

    test.todo('Test watching mode - online Players')
})

/*function generateGameSnapshot(game: any) {
    game.lastPlayed = 0;
    game.players.forEach((_: any, i: number) => game.players[i].name = `username${i}`)
    game.statistic.forEach((_: any, i: number) => game.statistic[i].actions.timePlayed = 0)
    return JSON.stringify(game)
}*/
