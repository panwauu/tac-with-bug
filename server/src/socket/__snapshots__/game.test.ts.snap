// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game test suite via socket.io > Test complete flow with all events > Post move 1`] = `"{"gamePlayer":2,"publicTournamentId":null,"privateTournamentId":null,"discardPile":["4","10","8","2","trickser","trickser","12","3","13","7","7"],"balls":[{"state":"locked","player":0,"position":83},{"state":"locked","player":0,"position":82},{"state":"locked","player":0,"position":80},{"state":"locked","player":0,"position":81},{"state":"locked","player":1,"position":86},{"state":"locked","player":1,"position":85},{"state":"locked","player":1,"position":84},{"state":"locked","player":1,"position":87},{"state":"locked","player":2,"position":89},{"state":"locked","player":2,"position":88},{"state":"locked","player":2,"position":91},{"state":"locked","player":2,"position":90},{"state":"locked","player":3,"position":93},{"state":"valid","player":3,"position":48},{"state":"locked","player":3,"position":95},{"state":"locked","player":3,"position":94}],"priorBalls":[{"state":"locked","player":0,"position":83},{"state":"locked","player":0,"position":82},{"state":"locked","player":0,"position":80},{"state":"locked","player":0,"position":81},{"state":"locked","player":1,"position":86},{"state":"locked","player":1,"position":85},{"state":"locked","player":1,"position":84},{"state":"locked","player":1,"position":87},{"state":"locked","player":2,"position":89},{"state":"locked","player":2,"position":88},{"state":"locked","player":2,"position":91},{"state":"locked","player":2,"position":90},{"state":"locked","player":3,"position":93},{"state":"valid","player":3,"position":41},{"state":"locked","player":3,"position":95},{"state":"locked","player":3,"position":94}],"cards":[{"title":"2","possible":false,"ballActions":{},"textAction":""},{"title":"3","possible":false,"ballActions":{},"textAction":""},{"title":"7","possible":false,"ballActions":{},"textAction":""}],"ownCards":["2","3","7"],"players":[{"name":"username0","remainingCards":2,"active":false,"playerNumber":0,"team":0,"narrFlag":[false,false],"discarded":false},{"name":"username1","remainingCards":2,"active":false,"playerNumber":1,"team":1,"narrFlag":[false,false],"discarded":true},{"name":"username2","remainingCards":3,"active":true,"playerNumber":2,"team":0,"narrFlag":[false,false],"discarded":false},{"name":"username3","remainingCards":2,"active":false,"playerNumber":3,"team":1,"narrFlag":[false,false],"discarded":false}],"gameEnded":true,"winningTeams":[true,false],"aussetzenFlag":false,"teufelFlag":false,"running":false,"coopCounter":-1,"statistic":[{"cards":{"1":[3,3,1],"4":[2,2,5],"7":[9,8,0],"8":[5,5,0],"13":[3,3,0],"tac":[2,2,0],"narr":[0,0,0],"engel":[0,0,0],"total":[83,75,17],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[2,2,1]},"actions":{"nMoves":67,"nAussetzen":4,"nBallsLost":0,"timePlayed":0,"nAbgeworfen":7,"nBallsKickedSelf":0,"nBallsKickedEnemy":0,"nBallsKickedOwnTeam":3}},{"cards":{"1":[6,5,1],"4":[2,2,0],"7":[4,4,0],"8":[4,4,0],"13":[4,4,2],"tac":[1,1,1],"narr":[0,0,0],"engel":[0,0,0],"total":[74,60,17],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[2,1,3]},"actions":{"nMoves":61,"nAussetzen":4,"nBallsLost":1,"timePlayed":0,"nAbgeworfen":12,"nBallsKickedSelf":0,"nBallsKickedEnemy":1,"nBallsKickedOwnTeam":1}},{"cards":{"1":[5,5,2],"4":[7,4,1],"7":[1,1,3],"8":[1,0,1],"13":[5,5,2],"tac":[3,2,0],"narr":[0,0,0],"engel":[0,0,0],"total":[79,59,19],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[3,2,0]},"actions":{"nMoves":64,"nAussetzen":0,"nBallsLost":7,"timePlayed":0,"nAbgeworfen":14,"nBallsKickedSelf":0,"nBallsKickedEnemy":3,"nBallsKickedOwnTeam":0}},{"cards":{"1":[3,3,1],"4":[3,3,1],"7":[0,0,1],"8":[3,2,2],"13":[6,6,0],"tac":[2,2,0],"narr":[0,0,0],"engel":[0,0,0],"total":[72,55,15],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[7,4,0]},"actions":{"nMoves":58,"nAussetzen":1,"nBallsLost":3,"timePlayed":0,"nAbgeworfen":13,"nBallsKickedSelf":0,"nBallsKickedEnemy":3,"nBallsKickedOwnTeam":0}}],"tradeDirection":-1,"deckInfo":[3,100],"colors":["red","blue","green","black"],"created":1611954094000,"lastPlayed":0,"rematch_open":true,"substitutedPlayerIndices":[],"substitution":null,"playernames":["UserA","UserB","UserC","UserD"],"teams":[[0,2],[1,3]]}"`;

exports[`Game test suite via socket.io > Test complete flow with all events > Post move 2`] = `"{"gamePlayer":3,"publicTournamentId":null,"privateTournamentId":null,"discardPile":["4","10","8","2","trickser","trickser","12","3","13","7","7"],"balls":[{"state":"locked","player":0,"position":83},{"state":"locked","player":0,"position":82},{"state":"locked","player":0,"position":80},{"state":"locked","player":0,"position":81},{"state":"locked","player":1,"position":86},{"state":"locked","player":1,"position":85},{"state":"locked","player":1,"position":84},{"state":"locked","player":1,"position":87},{"state":"locked","player":2,"position":89},{"state":"locked","player":2,"position":88},{"state":"locked","player":2,"position":91},{"state":"locked","player":2,"position":90},{"state":"locked","player":3,"position":93},{"state":"valid","player":3,"position":48},{"state":"locked","player":3,"position":95},{"state":"locked","player":3,"position":94}],"priorBalls":[{"state":"locked","player":0,"position":83},{"state":"locked","player":0,"position":82},{"state":"locked","player":0,"position":80},{"state":"locked","player":0,"position":81},{"state":"locked","player":1,"position":86},{"state":"locked","player":1,"position":85},{"state":"locked","player":1,"position":84},{"state":"locked","player":1,"position":87},{"state":"locked","player":2,"position":89},{"state":"locked","player":2,"position":88},{"state":"locked","player":2,"position":91},{"state":"locked","player":2,"position":90},{"state":"locked","player":3,"position":93},{"state":"valid","player":3,"position":41},{"state":"locked","player":3,"position":95},{"state":"locked","player":3,"position":94}],"cards":[{"title":"5","possible":false,"ballActions":{},"textAction":""},{"title":"7","possible":false,"ballActions":{},"textAction":""}],"ownCards":["5","7"],"players":[{"name":"username0","remainingCards":2,"active":false,"playerNumber":0,"team":0,"narrFlag":[false,false],"discarded":false},{"name":"username1","remainingCards":2,"active":false,"playerNumber":1,"team":1,"narrFlag":[false,false],"discarded":true},{"name":"username2","remainingCards":3,"active":true,"playerNumber":2,"team":0,"narrFlag":[false,false],"discarded":false},{"name":"username3","remainingCards":2,"active":false,"playerNumber":3,"team":1,"narrFlag":[false,false],"discarded":false}],"gameEnded":true,"winningTeams":[true,false],"aussetzenFlag":false,"teufelFlag":false,"running":false,"coopCounter":-1,"statistic":[{"cards":{"1":[3,3,1],"4":[2,2,5],"7":[9,8,0],"8":[5,5,0],"13":[3,3,0],"tac":[2,2,0],"narr":[0,0,0],"engel":[0,0,0],"total":[83,75,17],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[2,2,1]},"actions":{"nMoves":67,"nAussetzen":4,"nBallsLost":0,"timePlayed":0,"nAbgeworfen":7,"nBallsKickedSelf":0,"nBallsKickedEnemy":0,"nBallsKickedOwnTeam":3}},{"cards":{"1":[6,5,1],"4":[2,2,0],"7":[4,4,0],"8":[4,4,0],"13":[4,4,2],"tac":[1,1,1],"narr":[0,0,0],"engel":[0,0,0],"total":[74,60,17],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[2,1,3]},"actions":{"nMoves":61,"nAussetzen":4,"nBallsLost":1,"timePlayed":0,"nAbgeworfen":12,"nBallsKickedSelf":0,"nBallsKickedEnemy":1,"nBallsKickedOwnTeam":1}},{"cards":{"1":[5,5,2],"4":[7,4,1],"7":[1,1,3],"8":[1,0,1],"13":[5,5,2],"tac":[3,2,0],"narr":[0,0,0],"engel":[0,0,0],"total":[79,59,19],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[3,2,0]},"actions":{"nMoves":64,"nAussetzen":0,"nBallsLost":7,"timePlayed":0,"nAbgeworfen":14,"nBallsKickedSelf":0,"nBallsKickedEnemy":3,"nBallsKickedOwnTeam":0}},{"cards":{"1":[3,3,1],"4":[3,3,1],"7":[0,0,1],"8":[3,2,2],"13":[6,6,0],"tac":[2,2,0],"narr":[0,0,0],"engel":[0,0,0],"total":[72,55,15],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[7,4,0]},"actions":{"nMoves":58,"nAussetzen":1,"nBallsLost":3,"timePlayed":0,"nAbgeworfen":13,"nBallsKickedSelf":0,"nBallsKickedEnemy":3,"nBallsKickedOwnTeam":0}}],"tradeDirection":-1,"deckInfo":[3,100],"colors":["red","blue","green","black"],"created":1611954094000,"lastPlayed":0,"rematch_open":true,"substitutedPlayerIndices":[],"substitution":null,"playernames":["UserA","UserB","UserC","UserD"],"teams":[[0,2],[1,3]]}"`;

exports[`Game test suite via socket.io > Test complete flow with all events > Register first player 1`] = `"{"gamePlayer":2,"publicTournamentId":null,"privateTournamentId":null,"discardPile":["4","10","8","2","trickser","trickser","12","3","13","7","7"],"balls":[{"state":"locked","player":0,"position":83},{"state":"locked","player":0,"position":82},{"state":"locked","player":0,"position":80},{"state":"locked","player":0,"position":81},{"state":"locked","player":1,"position":86},{"state":"locked","player":1,"position":85},{"state":"locked","player":1,"position":84},{"state":"locked","player":1,"position":87},{"state":"locked","player":2,"position":89},{"state":"locked","player":2,"position":88},{"state":"locked","player":2,"position":91},{"state":"locked","player":2,"position":90},{"state":"locked","player":3,"position":93},{"state":"valid","player":3,"position":48},{"state":"locked","player":3,"position":95},{"state":"locked","player":3,"position":94}],"priorBalls":[{"state":"locked","player":0,"position":83},{"state":"locked","player":0,"position":82},{"state":"locked","player":0,"position":80},{"state":"locked","player":0,"position":81},{"state":"locked","player":1,"position":86},{"state":"locked","player":1,"position":85},{"state":"locked","player":1,"position":84},{"state":"locked","player":1,"position":87},{"state":"locked","player":2,"position":89},{"state":"locked","player":2,"position":88},{"state":"locked","player":2,"position":91},{"state":"locked","player":2,"position":90},{"state":"locked","player":3,"position":93},{"state":"valid","player":3,"position":41},{"state":"locked","player":3,"position":95},{"state":"locked","player":3,"position":94}],"cards":[{"title":"2","possible":true,"textAction":"beenden","ballActions":{}},{"title":"3","possible":true,"textAction":"beenden","ballActions":{}},{"title":"7","possible":true,"textAction":"beenden","ballActions":{}}],"ownCards":["2","3","7"],"players":[{"name":"username0","remainingCards":2,"active":false,"playerNumber":0,"team":0,"narrFlag":[false,false],"discarded":false},{"name":"username1","remainingCards":2,"active":false,"playerNumber":1,"team":1,"narrFlag":[false,false],"discarded":true},{"name":"username2","remainingCards":3,"active":true,"playerNumber":2,"team":0,"narrFlag":[false,false],"discarded":false},{"name":"username3","remainingCards":2,"active":false,"playerNumber":3,"team":1,"narrFlag":[false,false],"discarded":false}],"gameEnded":false,"winningTeams":[true,false],"aussetzenFlag":false,"teufelFlag":false,"running":true,"coopCounter":-1,"statistic":[{"cards":{"1":[3,3,1],"4":[2,2,5],"7":[9,8,0],"8":[5,5,0],"13":[3,3,0],"tac":[2,2,0],"narr":[0,0,0],"engel":[0,0,0],"total":[83,75,17],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[2,2,1]},"actions":{"nMoves":67,"nAussetzen":4,"nBallsLost":0,"timePlayed":0,"nAbgeworfen":7,"nBallsKickedSelf":0,"nBallsKickedEnemy":0,"nBallsKickedOwnTeam":3}},{"cards":{"1":[6,5,1],"4":[2,2,0],"7":[4,4,0],"8":[4,4,0],"13":[4,4,2],"tac":[1,1,1],"narr":[0,0,0],"engel":[0,0,0],"total":[74,60,17],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[2,1,3]},"actions":{"nMoves":61,"nAussetzen":4,"nBallsLost":1,"timePlayed":0,"nAbgeworfen":12,"nBallsKickedSelf":0,"nBallsKickedEnemy":1,"nBallsKickedOwnTeam":1}},{"cards":{"1":[5,5,2],"4":[7,4,1],"7":[1,1,3],"8":[1,0,1],"13":[5,5,2],"tac":[3,2,0],"narr":[0,0,0],"engel":[0,0,0],"total":[78,58,19],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[3,2,0]},"actions":{"nMoves":63,"nAussetzen":0,"nBallsLost":7,"timePlayed":0,"nAbgeworfen":14,"nBallsKickedSelf":0,"nBallsKickedEnemy":3,"nBallsKickedOwnTeam":0}},{"cards":{"1":[3,3,1],"4":[3,3,1],"7":[0,0,1],"8":[3,2,2],"13":[6,6,0],"tac":[2,2,0],"narr":[0,0,0],"engel":[0,0,0],"total":[72,55,15],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[7,4,0]},"actions":{"nMoves":58,"nAussetzen":1,"nBallsLost":3,"timePlayed":0,"nAbgeworfen":13,"nBallsKickedSelf":0,"nBallsKickedEnemy":3,"nBallsKickedOwnTeam":0}}],"tradeDirection":-1,"deckInfo":[3,100],"colors":["red","blue","green","black"],"created":1611954094000,"lastPlayed":0,"rematch_open":false,"substitutedPlayerIndices":[],"substitution":null,"playernames":["UserA","UserB","UserC","UserD"],"teams":[[0,2],[1,3]]}"`;

exports[`Game test suite via socket.io > Test complete flow with all events > Register second player 1`] = `"{"gamePlayer":3,"publicTournamentId":null,"privateTournamentId":null,"discardPile":["4","10","8","2","trickser","trickser","12","3","13","7","7"],"balls":[{"state":"locked","player":0,"position":83},{"state":"locked","player":0,"position":82},{"state":"locked","player":0,"position":80},{"state":"locked","player":0,"position":81},{"state":"locked","player":1,"position":86},{"state":"locked","player":1,"position":85},{"state":"locked","player":1,"position":84},{"state":"locked","player":1,"position":87},{"state":"locked","player":2,"position":89},{"state":"locked","player":2,"position":88},{"state":"locked","player":2,"position":91},{"state":"locked","player":2,"position":90},{"state":"locked","player":3,"position":93},{"state":"valid","player":3,"position":48},{"state":"locked","player":3,"position":95},{"state":"locked","player":3,"position":94}],"priorBalls":[{"state":"locked","player":0,"position":83},{"state":"locked","player":0,"position":82},{"state":"locked","player":0,"position":80},{"state":"locked","player":0,"position":81},{"state":"locked","player":1,"position":86},{"state":"locked","player":1,"position":85},{"state":"locked","player":1,"position":84},{"state":"locked","player":1,"position":87},{"state":"locked","player":2,"position":89},{"state":"locked","player":2,"position":88},{"state":"locked","player":2,"position":91},{"state":"locked","player":2,"position":90},{"state":"locked","player":3,"position":93},{"state":"valid","player":3,"position":41},{"state":"locked","player":3,"position":95},{"state":"locked","player":3,"position":94}],"cards":[{"title":"5","possible":false,"ballActions":{},"textAction":""},{"title":"7","possible":false,"ballActions":{},"textAction":""}],"ownCards":["5","7"],"players":[{"name":"username0","remainingCards":2,"active":false,"playerNumber":0,"team":0,"narrFlag":[false,false],"discarded":false},{"name":"username1","remainingCards":2,"active":false,"playerNumber":1,"team":1,"narrFlag":[false,false],"discarded":true},{"name":"username2","remainingCards":3,"active":true,"playerNumber":2,"team":0,"narrFlag":[false,false],"discarded":false},{"name":"username3","remainingCards":2,"active":false,"playerNumber":3,"team":1,"narrFlag":[false,false],"discarded":false}],"gameEnded":false,"winningTeams":[true,false],"aussetzenFlag":false,"teufelFlag":false,"running":true,"coopCounter":-1,"statistic":[{"cards":{"1":[3,3,1],"4":[2,2,5],"7":[9,8,0],"8":[5,5,0],"13":[3,3,0],"tac":[2,2,0],"narr":[0,0,0],"engel":[0,0,0],"total":[83,75,17],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[2,2,1]},"actions":{"nMoves":67,"nAussetzen":4,"nBallsLost":0,"timePlayed":0,"nAbgeworfen":7,"nBallsKickedSelf":0,"nBallsKickedEnemy":0,"nBallsKickedOwnTeam":3}},{"cards":{"1":[6,5,1],"4":[2,2,0],"7":[4,4,0],"8":[4,4,0],"13":[4,4,2],"tac":[1,1,1],"narr":[0,0,0],"engel":[0,0,0],"total":[74,60,17],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[2,1,3]},"actions":{"nMoves":61,"nAussetzen":4,"nBallsLost":1,"timePlayed":0,"nAbgeworfen":12,"nBallsKickedSelf":0,"nBallsKickedEnemy":1,"nBallsKickedOwnTeam":1}},{"cards":{"1":[5,5,2],"4":[7,4,1],"7":[1,1,3],"8":[1,0,1],"13":[5,5,2],"tac":[3,2,0],"narr":[0,0,0],"engel":[0,0,0],"total":[78,58,19],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[3,2,0]},"actions":{"nMoves":63,"nAussetzen":0,"nBallsLost":7,"timePlayed":0,"nAbgeworfen":14,"nBallsKickedSelf":0,"nBallsKickedEnemy":3,"nBallsKickedOwnTeam":0}},{"cards":{"1":[3,3,1],"4":[3,3,1],"7":[0,0,1],"8":[3,2,2],"13":[6,6,0],"tac":[2,2,0],"narr":[0,0,0],"engel":[0,0,0],"total":[72,55,15],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[7,4,0]},"actions":{"nMoves":58,"nAussetzen":1,"nBallsLost":3,"timePlayed":0,"nAbgeworfen":13,"nBallsKickedSelf":0,"nBallsKickedEnemy":3,"nBallsKickedOwnTeam":0}}],"tradeDirection":-1,"deckInfo":[3,100],"colors":["red","blue","green","black"],"created":1611954094000,"lastPlayed":0,"rematch_open":false,"substitutedPlayerIndices":[],"substitution":null,"playernames":["UserA","UserB","UserC","UserD"],"teams":[[0,2],[1,3]]}"`;

exports[`Game test suite via socket.io > Test dealCards > Register player and expect dealCards 1`] = `"{"gamePlayer":0,"publicTournamentId":null,"privateTournamentId":null,"discardPile":[],"balls":[{"state":"locked","player":0,"position":83},{"state":"locked","player":0,"position":82},{"state":"locked","player":0,"position":80},{"state":"locked","player":0,"position":81},{"state":"locked","player":1,"position":86},{"state":"locked","player":1,"position":85},{"state":"locked","player":1,"position":84},{"state":"locked","player":1,"position":87},{"state":"locked","player":2,"position":89},{"state":"locked","player":2,"position":88},{"state":"locked","player":2,"position":91},{"state":"locked","player":2,"position":90},{"state":"locked","player":3,"position":93},{"state":"valid","player":3,"position":48},{"state":"locked","player":3,"position":95},{"state":"locked","player":3,"position":94}],"priorBalls":[{"state":"locked","player":0,"position":83},{"state":"locked","player":0,"position":82},{"state":"locked","player":0,"position":80},{"state":"locked","player":0,"position":81},{"state":"locked","player":1,"position":86},{"state":"locked","player":1,"position":85},{"state":"locked","player":1,"position":84},{"state":"locked","player":1,"position":87},{"state":"locked","player":2,"position":89},{"state":"locked","player":2,"position":88},{"state":"locked","player":2,"position":91},{"state":"locked","player":2,"position":90},{"state":"locked","player":3,"position":93},{"state":"valid","player":3,"position":48},{"state":"locked","player":3,"position":95},{"state":"locked","player":3,"position":94}],"cards":[{"title":"3","possible":true,"ballActions":{},"textAction":"tauschen"},{"title":"6","possible":true,"ballActions":{},"textAction":"tauschen"},{"title":"tac","possible":true,"ballActions":{},"textAction":"tauschen"},{"title":"7","possible":true,"ballActions":{},"textAction":"tauschen"},{"title":"4","possible":true,"ballActions":{},"textAction":"tauschen"}],"ownCards":["3","6","tac","7","4"],"players":[{"name":"username0","remainingCards":5,"active":true,"playerNumber":0,"team":0,"narrFlag":[false,false],"discarded":false,"tradeInformation":[false,false]},{"name":"username1","remainingCards":5,"active":false,"playerNumber":1,"team":1,"narrFlag":[false,false],"discarded":true,"tradeInformation":[true,false]},{"name":"username2","remainingCards":5,"active":false,"playerNumber":2,"team":0,"narrFlag":[false,false],"discarded":false,"tradeInformation":[true,false]},{"name":"username3","remainingCards":5,"active":false,"playerNumber":3,"team":1,"narrFlag":[false,false],"discarded":false,"tradeInformation":[true,false]}],"gameEnded":false,"winningTeams":[true,false],"aussetzenFlag":false,"teufelFlag":false,"running":true,"coopCounter":-1,"statistic":[{"cards":{"1":[3,3,1],"4":[2,2,5],"7":[9,8,0],"8":[5,5,0],"13":[3,3,0],"tac":[2,2,0],"narr":[0,0,0],"engel":[0,0,0],"total":[83,75,17],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[2,2,1]},"actions":{"nMoves":67,"nAussetzen":4,"nBallsLost":0,"timePlayed":0,"nAbgeworfen":7,"nBallsKickedSelf":0,"nBallsKickedEnemy":0,"nBallsKickedOwnTeam":3}},{"cards":{"1":[6,5,1],"4":[2,2,0],"7":[4,4,0],"8":[4,4,0],"13":[4,4,2],"tac":[1,1,1],"narr":[0,0,0],"engel":[0,0,0],"total":[74,60,17],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[2,1,3]},"actions":{"nMoves":61,"nAussetzen":4,"nBallsLost":1,"timePlayed":0,"nAbgeworfen":12,"nBallsKickedSelf":0,"nBallsKickedEnemy":1,"nBallsKickedOwnTeam":1}},{"cards":{"1":[5,5,2],"4":[7,4,1],"7":[1,1,3],"8":[1,0,1],"13":[5,5,2],"tac":[3,2,0],"narr":[0,0,0],"engel":[0,0,0],"total":[78,58,19],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[3,2,0]},"actions":{"nMoves":63,"nAussetzen":0,"nBallsLost":7,"timePlayed":0,"nAbgeworfen":14,"nBallsKickedSelf":0,"nBallsKickedEnemy":3,"nBallsKickedOwnTeam":0}},{"cards":{"1":[3,3,1],"4":[3,3,1],"7":[0,0,1],"8":[3,2,2],"13":[6,6,0],"tac":[2,2,0],"narr":[0,0,0],"engel":[0,0,0],"total":[72,55,15],"teufel":[0,0,0],"krieger":[0,0,0],"trickser":[7,4,0]},"actions":{"nMoves":58,"nAussetzen":1,"nBallsLost":3,"timePlayed":0,"nAbgeworfen":13,"nBallsKickedSelf":0,"nBallsKickedEnemy":3,"nBallsKickedOwnTeam":0}}],"tradeDirection":-1,"deckInfo":[0,80],"colors":["red","blue","green","black"],"created":1611954094000,"lastPlayed":0,"rematch_open":false,"discardedFlag":false,"substitutedPlayerIndices":[],"substitution":null,"playernames":["UserA","UserB","UserC","UserD"],"teams":[[0,2],[1,3]]}"`;
